bindings/serialize.h

```diff
@@ -0,0 +1,185 @@
+  virtual void encode(std::vector<char> &b);
+};
+
+
+// Corresponds to Mapping defined in third_party/proto/profile.proto.
+class Mapping : public ProtoField {
+private:
+  uint64_t id;
+  uint64_t start;
+  uint64_t limit;
+  uint64_t offset;
+  uint64_t fileX; // Index into string table.  
+  uint64_t buildIDX; // Index into string table.  
+  bool hasFunctions;
+  bool hasFilenames;
+  bool hasLineNumbers;
+  bool hasInlineFrames;
+
+public:
+  Mapping(uint64_t id, uint64_t start, uint64_t limit, uint64_t offset,
+          uint64_t fileX, uint64_t buildIDX, bool hasFunctions,
+          bool hasFilenames, bool hasLineNumbers, bool hasInlineFrames);
+  virtual void encode(std::vector<char> &b);
+};
+
+// Corresponds to Line defined in third_party/proto/profile.proto.
+class Line : public ProtoField {
+private:
+  uint64_t functionID;
+  int64_t line;
+
+public:
+  Line(uint64_t functionID, int64_t line);
+  virtual void encode(std::vector<char> &b);
+};
+
+// Corresponds to Function defined in third_party/proto/profile.proto.
+class ProfileFunction : public ProtoField {
+private:
+  uint64_t id;
+  int64_t nameX; // Index into string table. 
+  int64_t systemNameX; // Index into string table. 
+  int64_t filenameX; // Index into string table. 
+  int64_t startLine;
+
+public:
+  ProfileFunction(uint64_t id, int64_t nameX, int64_t systemNameX,
+                  int64_t filenameX, int64_t startLine);
+  virtual void encode(std::vector<char> &b);
+};
+
+// Corresponds to Location defined in third_party/proto/profile.proto.
+class ProfileLocation : public ProtoField {
+private:
+  uint64_t id;
+  uint64_t mappingID;
+  uint64_t address;
+  std::vector<Line> line;
+  bool isFolded;
+
+public:
+  ProfileLocation(uint64_t id, uint64_t mappingID, uint64_t address,
+                  std::vector<Line> line, bool isFolded);
+  virtual void encode(std::vector<char> &b);
+};
+
+// Corresponds to Sample defined in third_party/proto/profile.proto.
+class Sample : public ProtoField {
+private:
+  std::vector<uint64_t> locationID;
+  std::vector<int64_t> value;
+  std::vector<Label> label;
+
+public:
+  Sample(std::vector<uint64_t> locationID, std::vector<int64_t> value,
+         std::vector<Label> label);
+  virtual void encode(std::vector<char> &b);
+};
+
+class Node;
+
+// Corresponds to Profile defined in third_party/proto/profile.proto.
+class Profile : public ProtoField {
+private:
+  std::vector<ValueType> sampleType;
+  std::vector<ProfileLocation> location;
+  std::vector<Sample> sample;
+  std::vector<Mapping> mapping;
+  std::vector<ProfileFunction> function;
+  std::vector<std::string> strings;
+  int64_t dropFramesX; // Index into string table. 
+  int64_t keepFramesX; // Index into string table. 
+  int64_t timeNanos;
+  int64_t durationNanos;
+  ValueType periodType;
+  int64_t period;
+  std::vector<int64_t> commentX; // Indices into string table. 
+  int64_t defaultSampleTypeX;
+  std::map<std::string, int64_t> functionIdMap;
```

Comment:
If you don't care about the order, use unordered_map, that uses hash map rather than rb tree underneath which is more efficient (amortized O(1) for insertions and lookup).

Reply:
Done.